modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/Commons.java

```diff
@@ -598,12 +601,12 @@ public static int nextPowerOf2(int v) {
      * @param elems Elems.
      * @return The longest possible prefix of {@code seq}.
      */
-    public static <T> List<T> maxPrefix(List<T> seq, Collection<T> elems) {
-        List<T> res = new ArrayList<>();
+    public static IntList maxPrefix(List<Integer> seq, Collection<Integer> elems) {
```

Comment:
seems you could use ImutableIntList intead of List<Integer> here + rewrite iteration through the collection

Reply:
done